### DISCUSSION TITLE
Creating Jupyter notebooks for exercises & solutions

### DISCUSSION THREAD

**Discussion by tomsutch** [quarto-dev/quarto-cli#2422] [Q&A]
Hi, I'm thinking of using Quarto to deliver a workshop, producing slides, notes and Jupyter notebooks from a common set of source documents. As part of this I'll be preparing exercises for attendees to attempt using Python code in Jupyter notebooks - and then separately providing worked solutions for these exercises. I've got the following workflow working:

1. Author the content in a qmd file - say `my-exercise.qmd`. For chunks that I want attendees to try themselves, include `#| tags: [solution]` in the chunk options.
2. Convert the qmd to a Jupyter notebook using `quarto convert my-exercise.qmd -o my-solution-notebook.ipynb`. (This will be the solution notebook.)
3. Use a Python script, using `nbformat`, to create the exercise notebook from the solution notebook.

```python
import nbformat
nb = nbformat.read("my-solution-notebook.ipynb", as_version=4)
for cell in nb.cells:
    if cell.cell_type == 'code':
        tags = cell.metadata.get('tags', list())
        if 'solution' in tags:
            cell.source = ""
            cell['outputs'] = []
            cell.execution_count = None
nbformat.write(nb, "my-exercise-notebook.ipynb")
```

So it works, if a bit messily (I can make the Python a bit neater by turning it into a script accepting command-line arguments). Is this the best way of doing it, or is there a way of filtering the 'solution' chunks out using Quarto, perhaps at the `quarto convert` stage?

Thanks for your help!

**Comment by lexnederbragt** [comment]
I am experimenting with the same. It looks like this can be achieved using [project profiles](https://quarto.org/docs/projects/profiles.html).

A working example. `my-exercise.qmd`:

```
Exercise title and description


::: {.content-visible when-profile="solutions"}

Solution

:::
```

To generate the exercise notebook:

```
quarto render my-exercise.qmd --to ipynb -o my-exercise-notebook.ipynb
```

To generate the solution notebook:

```
quarto render my-exercise.qmd --to ipynb --profile solutions -o my-solution-notebook.ipynb
```

By the way, I combine this with the `::: {.cell .markdown}` div (https://pandoc.org/MANUAL.html#jupyter-notebooks) so learners, if the notebook contains multiple exercises, can add a code cell to write their solution in below the exercise (hope that makes sense).
